### PR TITLE
Build autoscaler abstractions like fetcher, client and scaler

### PR DESCRIPTION
**Pull request by Jeffwan**
## Pull Request Description
This pull request includes several significant updates to the Pod Autoscaling algorithms and their related components. The changes introduce new scaling algorithms, enhance the existing ones, and update the metrics fetching mechanisms. Below are the most important changes grouped by theme.

### New and Enhanced Scaling Algorithms:

* Introduced the `ScalingAlgorithm` interface and extract the `ApaScalingAlgorithm` and `KpaScalingAlgorithm` classes with their respective `ComputeTargetReplicas` methods.
* Refactored the metrics client to use a new `PodMetricClient` structure and introduced methods for updating and fetching pod metrics. 
* Added a new `MetricFetcher` interface and implemented it with `RestMetricsFetcher`, `ResourceMetricsFetcher`, and `CustomMetricsFetcher` classes. (`pkg/controller/podautoscaler/metrics/fetcher.go`)

These changes collectively enhance the autoscaling capabilities and metrics handling of the Pod Autoscaler, making it more robust and flexible for different scaling scenarios.

## Related Issues
Resolves: part of https://github.com/aibrix/aibrix/issues/119

**Important: Before submitting, please complete the description above and review the checklist below.**

---

<details>
<summary><strong>Contribution Guidelines (Expand for Details)</strong></summary>

<p>We appreciate your contribution to aibrix! To ensure a smooth review process and maintain high code quality, please adhere to the following guidelines:</p>

<h3>Pull Request Title Format</h3>
<p>Your PR title should start with one of these prefixes to indicate the nature of the change:</p>
<ul>
    <li><code>[Bug]</code>: Corrections to existing functionality</li>
    <li><code>[CI]</code>: Changes to build process or CI pipeline</li>
    <li><code>[Docs]</code>: Updates or additions to documentation</li>
    <li><code>[API]</code>: Modifications to aibrix's API or interface</li>
    <li><code>[CLI]</code>: Changes or additions to the Command Line Interface</li>
    <li><code>[Misc]</code>: For changes not covered above (use sparingly)</li>
</ul>
<p><em>Note: For changes spanning multiple categories, use multiple prefixes in order of importance.</em></p>

<h3>Submission Checklist</h3>
<ul>
    <li>[ ] PR title includes appropriate prefix(es)</li>
    <li>[ ] Changes are clearly explained in the PR description</li>
    <li>[ ] New and existing tests pass successfully</li>
    <li>[ ] Code adheres to project style and best practices</li>
    <li>[ ] Documentation updated to reflect changes (if applicable)</li>
    <li>[ ] Thorough testing completed, no regressions introduced</li>
</ul>

<p>By submitting this PR, you confirm that you've read these guidelines and your changes align with the project's contribution standards.</p>

</details>

### CODE REVIEW

File: pkg/controller/podautoscaler/metrics/client.go [hunk 1]
```diff
@@ -38,26 +37,38 @@ const (
...
-	//TODO implement me
+func (c *PodMetricClient) UpdatePodListMetric(ctx context.Context, metricKey NamespaceNameMetric, podList *corev1.PodList, containerPort int, now time.Time) error {
 	panic("implement me")
 }
 
+func NewMetricsClient(fetcher MetricFetcher) *PodMetricClient {
+	return &PodMetricClient{fetcher: fetcher}
+}
+
+func (c *PodMetricClient) UpdatePodMetrics(ctx context.Context, podList []v1.Pod, containerPort int, metricName string) ([]float64, error) {
```
Comment (kr11) [inline pkg/controller/podautoscaler/metrics/client.go]: Maybe I miss it, where do we call the `UpdatePodMetrics` and perform periodic updates?

File: pkg/controller/podautoscaler/algorithm/kpa.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	"math"
+
+	"github.com/aibrix/aibrix/pkg/controller/podautoscaler/common"
+)
+
+type KpaScalingAlgorithm struct{}
+
+var _ ScalingAlgorithm = (*KpaScalingAlgorithm)(nil)
+
+func (a *KpaScalingAlgorithm) ComputeTargetReplicas(currentPodCount float64, context common.ScalingContext) int32 {
```
Comment (kr11) [inline pkg/controller/podautoscaler/algorithm/kpa.go:29]: what's the relationship between this `ComputeTargetReplicas` and `(k *KpaAutoscaler) Scale` in `controller/podautoscaler/scaler/kpa.go`?  How will these two parts collaborate?

File: pkg/controller/podautoscaler/metrics/client.go (hunk 1 above)
Comment (Jeffwan) [reply pkg/controller/podautoscaler/metrics/client.go]: no reference yet. I think it should be `GetMetricsFromPods`, I make a new method but didn't delete the original one for comparision. I should rename it back to `GetMetricsFromPods` and delete this method.

File: pkg/controller/podautoscaler/algorithm/kpa.go (hunk 2 above)
Comment (Jeffwan) [reply pkg/controller/podautoscaler/algorithm/kpa.go:29]: Current abstraction should work in this way.

1. algorithm should be able to calculate the target replicas without any further dependencies. it should be very clean stateless logics
2. scaler should orchestrate everything needed for autoscaling, like get metrics using the client,  calculate the right replica, make changes to deployment etc.

**Pull request merged**